app/models/bitbucket_integration.rb

```diff
@@ -0,0 +1,243 @@
+# == Schema Information
+#
+# Table name: bitbucket_integrations
+#
+#  id                  :uuid             not null, primary key
+#  oauth_access_token  :string
+#  oauth_refresh_token :string
+#  created_at          :datetime         not null
+#  updated_at          :datetime         not null
+#
+class BitbucketIntegration < ApplicationRecord
+  has_paper_trail
+  include Vaultable
+  include Providable
+  include Displayable
+  include Rails.application.routes.url_helpers
+
+  encrypts :oauth_access_token, deterministic: true
+  encrypts :oauth_refresh_token, deterministic: true
+
+  BASE_INSTALLATION_URL =
+    Addressable::Template.new("https://bitbucket.org/site/oauth2/authorize{?params*}")
+
+  attr_accessor :code
+  before_create :complete_access
+  delegate :code_repo_name_only, to: :app_config
+
+  def install_path
+    unless integration.version_control? || integration.ci_cd?
+      raise Integration::IntegrationNotImplemented, "We don't support that yet!"
+    end
+
+    BASE_INSTALLATION_URL
+      .expand(params: {
+        client_id: creds.integrations.bitbucket.client_id,
+        redirect_uri: redirect_uri,
+        response_type: :code,
+        state: integration.installation_state
+      }).to_s
+  end
+
+  def complete_access
+    set_tokens(Installations::Bitbucket::Api.oauth_access_token(code, redirect_uri))
+  end
+
+  def installation
+    Installations::Bitbucket::Api.new(oauth_access_token, workspace)
+  end
+
+  def to_s
+    "bitbucket"
+  end
+
+  def creatable?
+    false
+  end
+
+  def connectable?
+    true
+  end
+
+  def connection_data
+    return unless integration.metadata
+    "Organization: #{integration.metadata["account_name"]} (#{integration.metadata["account_id"]})"
+  end
+
+  def further_setup?
+    return true if integration.version_control?
+    false
+  end
+
+  WORKSPACE_TRANSFORMATIONS = {
+    id: :slug,
+    name: :name
+  }
+
+  def workspaces
+    with_api_retries { installation.list_workspaces(WORKSPACE_TRANSFORMATIONS) }
+  end
+
+  # VCS
+
+  REPOS_TRANSFORMATIONS = {
+    id: :slug,
+    name: :name,
+    namespace: [:workspace, :slug],
+    full_name: :full_name,
+    description: :description,
+    repo_url: [:links, :html],
+    avatar_url: [:links, :avatar]
+  }
+
+  WEBHOOK_TRANSFORMATIONS = {
+    id: :uuid,
+    url: :url,
+    events: :events
+  }
+
+  PR_TRANSFORMATIONS = {
+    source_id: :id,
+    number: :id,
+    title: :title,
+    body: :description,
+    url: [:links, :html, :href],
+    state: :state,
+    head_ref: [:source, :branch, :name],
+    base_ref: [:destination, :branch, :name],
+    opened_at: :created_on
+  }
+
+  COMMITS_TRANSFORMATIONS = {
+    url: [:links, :html, :href],
+    commit_hash: :hash,
+    message: :message,
+    # TODO: author information is not available when user does
+    # not match on atlassian
+    author_name: [:author, :user, :display_name],
+    # TODO: email is not available
```

Comment:
Bitbucket doesn't provide the author email, should we instead make use of `[:author :raw] => Brodie Rao <a@b.c>`?